analyzers/src/SonarAnalyzer.SourceGenerator/RuleCatalogGenerator.cs

```diff
@@ -0,0 +1,106 @@
+﻿/*
+ * SonarAnalyzer for .NET
+ * Copyright (C) 2015-2022 SonarSource SA
+ * mailto: contact AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics.CodeAnalysis;
+using System.IO;
+using System.Linq;
+using System.Text;
+using Microsoft.CodeAnalysis;
+using Newtonsoft.Json.Linq;
+
+namespace SonarAnalyzer.SourceGenerator
+{
+    [Generator]
+    [ExcludeFromCodeCoverage]
+    public class RuleCatalogGenerator : ISourceGenerator
+    {
+        private const string SonarWayFileName = "Sonar_way_profile.json";
+
+        public void Initialize(GeneratorInitializationContext context)
+        {
+            // Not needed
+        }
+
+        public void Execute(GeneratorExecutionContext context)
+        {
+            if (!context.AnalyzerConfigOptions.GlobalOptions.TryGetValue("build_property.projectdir", out var projectDir))
+            {
+                throw new NotSupportedException("Cannot find ProjectDir");
+            }
+            var language = Path.GetFileName(projectDir.TrimEnd(Path.DirectorySeparatorChar)) switch
+            {
+                "SonarAnalyzer.CSharp" => "cs",
+                "SonarAnalyzer.VisualBasic" => "vbnet",
+                _ => throw new ArgumentException($"Unexpected projectDir: {projectDir}")
+            };
+            context.AddSource($"RuleCatalog.{language}.g.cs", GenerateSource(RuleDescriptorArguments(Path.Combine(projectDir, "..", "..", "rspec", language))));
+        }
+
+        private static IEnumerable<string[]> RuleDescriptorArguments(string rspecDirectory)
+        {
+            var sonarWay = new HashSet<string>(JObject.Parse(File.ReadAllText(Path.Combine(rspecDirectory, SonarWayFileName)))["ruleKeys"].Values<string>());
+            foreach (var jsonPath in Directory.GetFiles(rspecDirectory, "*.json").Where(x => Path.GetFileName(x) != SonarWayFileName))
+            {
+                var json = JObject.Parse(File.ReadAllText(jsonPath));
+                var id = Path.GetFileName(jsonPath).Split('_').First();
+                yield return new[]
```

Comment:
I'll need to add few UTs with data-dependent assertions for specific rules. To visually demonstrate the values for couple of rules (one hotspot, one bug, one code smell...)

Reply:
Added